sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/trees/TreeNode.scala

```diff
@@ -72,6 +72,34 @@ object CurrentOrigin {
   }
 }
 
+case class Barrier(node: Option[TreeNode[_]] = None)
```

Comment:
My original thought is: If we use a barrier node, we need to modify many places where we create a new logical plan and wrap it with the barrier node.

I will revamp it with a barrier node.